apps/web/src/hooks/useRequiredParams.ts

```diff
@@ -0,0 +1,51 @@
+import invariant from "tiny-invariant";
+import { Params, useParams } from "react-router-dom";
+
+import { notEmpty } from "@gc-digital-talent/helpers";
+
+export const uuidRegEx =
+  /[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}/;
+
+export const isUUID = (str: string): boolean => {
+  return !!str.match(uuidRegEx);
+};
+
+const assertParam = (param?: string, enforceUUID: boolean = true) => {
+  invariant(
+    notEmpty(param),
+    `Could not find required URL parameter "${param}"`,
+  );
+
+  if (enforceUUID) {
+    invariant(isUUID(param), `URL parameter "${param}" must be a UUID.`);
+  }
+};
+
+type Keys<T, K extends keyof T> = K | Array<K>;
+
+const useRequiredParams = <
+  T extends Record<
+    keyof {
+      [P in keyof T as T[P] extends string ? P : never]: string | undefined;
```

Comment:
Wow, this is some dark TS voodoo.  Could you add some comments to help me out?

Reply:
Hopefully this makes sense 😆 

[document generic useRequiredParams](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8300/commits/be2c7572fd40d2026aaf8e5661ada63e672f5ba6)